tools/crm_resource_runtime.c

```diff
@@ -662,6 +662,24 @@ cli_resource_delete_failures(crm_ipc_t *crmd_channel, const char *host_uname,
 {
     int rc = pcmk_ok;
 
+    if (rsc == NULL) {
+        return -ENXIO;
+
+    } else if (rsc->children) {
+        GListPtr lpc = NULL;
+
+        for (lpc = rsc->children; lpc != NULL; lpc = lpc->next) {
```

Comment:
I don't think we need this loop, since cli_resource_delete() already loops through the rsc's children.

I suspect the problem is that we call cli_resource_delete() with a NULL rsc if the user didn't specify a rsc_id. Instead, we should probably do a pe_find_resource_with_flags() for resource_name before calling cli_resource_delete().

Reply:
> I suspect the problem is that we call cli_resource_delete() with a NULL rsc if the user didn't specify a rsc_id. Instead, we should probably do a pe_find_resource_with_flags() for resource_name before calling cli_resource_delete().

Ah, that's actually a different bug. :-)

I think the root of your issue is that the comparison:

     } else if(rsc->id && safe_str_neq(rsc->id, resource_name)) {

won't match when resource_name is a child of rsc. For that, we probably need a new function to compare a name against a resource and all of its descendants. Plus my fix suggested above.

Reply:
I think it makes sense only for primitives to go into this block and be checked against data_set->failed, that's the reason why we need the previous loop. 
Only the failed primitives should run into cli_resource_delete() and actually get cleaned up, and it'll simply skip the loop in there,

Reply:
Sure, that would work too. The above would become:

    if (rsc && rsc->children) {

Reply:
I'll handle the above

Reply:
Thanks, I was too busy to handle this today :-\ There are more things though. crm_resource --cleanup without specifying resource had been broken before this pull request actually.